File: Clojure/src/clojure/pi_quadrature/pi_sequential_map_reduce.clj
;  Calculation of π using quadrature. Sequential algorithm.
;
;  Copyright © 2013, 2014  Russel Winder

(ns pi-quadrature.pi-sequential-map-reduce)

(load "output")

(defn calculate [delta i]
  (let [x (* (- i 0.5) delta)]
    (/ 1.0 (+ 1.0 (* x x)))))

(defn summation [n delta]
   (reduce + (map (partial calculate delta) (range n))))

(defn -main []
  (let [
        n  10000000 ; 100 times fewer due to speed issues.
        delta  (/ 1.0 n)
        startTimeNanos  (System/nanoTime)
        pi  (* 4.0 delta (summation n delta))
        elapseTime  (/ (- (System/nanoTime) startTimeNanos) 1e9)
        ]
    (out "Sequential_Map_Reduce" pi n elapseTime)))

File: Kotlin/src/pi_sequential_map_reduce.kt
/*
 *  Calculation of π using quadrature realized with map and reduce.
 *
 *  Copyright © 2013–2015  Russel Winder
 */

package uk.org.winder.pi_quadrature.pi_sequential_map_reduce

import uk.org.winder.pi_quadrature.out

fun main(args:Array<String>) {
  val n = 1000000000
  val delta = 1.0 / n
  val startTimeNanos = System.nanoTime()
  val pi = 4.0 * delta * (1..n).sequence().map{i ->
    val x = (i - 0.5) * delta
    1.0 / (1.0 + x * x)
  }.reduce{t, i -> t + i}
  val elapseTime = (System.nanoTime() - startTimeNanos) / 1e9
  out("pi_sequential_map_reduce", pi, n, elapseTime)
}
